### ISSUE TITLE
Change return type to `Stream<? extends TestTemplateInvocationContext>`

### ISSUE THREAD

**Issue by sormuras** [junit-team/junit5#3577]
https://github.com/junit-team/junit5/blob/f5b78f68c6a90dfa63611c7a10027581c5d913ba/junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestTemplateInvocationContextProvider.java#L87

Michael wrote:
> I mean I tried to do `Stream.generate(() -> new MyInvocationContext()).limit(/*number*/)`
> This does not work.
> but `Stream.generate(() -> new MyInvocationContext()).limit(/*number*/).map(e -> e)` works...

**Comment by Michael1993** [comment]
I think the return value should be `Stream<? extends TestTemplateInvocationContext>`? I might be missing something.

**Comment by sormuras** [comment]
Question is, can we change the return type without breaking existing code?

**Comment by marcphilipp** [comment]
According to https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_interfaces_-_API_methods that would break binary compatibility. We should have declared the method like that initially, though.

**Comment by kaipe** [comment]
On a related topic ...
Wouldn't it be better to have return type Iterable?
I figure it's a simplier abstraction that is easy to implement and still provides all capabilities that are requied by JUnit.

**Comment by marcphilipp** [comment]
I don't think it makes a big difference. `Stream` nowadays has many factory methods that allow `Iterator`-like implementations. Moreover, we ensure `Stream.close` is also called to support external resources such as files.
